Sources/MongoSwift/ChangeStream.swift

```diff
@@ -90,60 +105,101 @@ public class ChangeStream<T: Codable>: Sequence, IteratorProtocol {
+        guard case let .open(changeStreamPtr, _, _) = self.state else {
+            throw ClosedChangeStreamError
+        }
+
+        let out = UnsafeMutablePointer<BSONPointer?>.allocate(capacity: 1)
+        defer {
+            out.deinitialize(count: 1)
+            out.deallocate()
+        }
+
+        guard mongoc_change_stream_next(changeStreamPtr, out) else {
+            if let error = self.getChangeStreamError() {
+                self.blockingClose()
+                throw error
+            }
+            return nil
+        }
+        guard let pointee = out.pointee else {
+            fatalError("The change stream was advanced, but the document is nil")
+        }
+
+        // We have to copy because libmongoc owns the pointer.
+        let doc = Document(copying: pointee)
+
+        // Update the resumeToken with the `_id` field from the document.
+        guard let resumeToken = doc["_id"]?.documentValue else {
+            throw InternalError(message: "_id field is missing from the change stream document.")
+        }
+        self.resumeToken = ResumeToken(resumeToken)
+        return try self.decoder.decode(T.self, from: doc)
+    }
+
+    /// Destroys the underlying mongoc change stream and closes the state in a blocking manner.
+    private func blockingClose() {
+        guard case let .open(changeStream, _, _) = self.state else {
+            return
+        }
+        self.pollingSemaphore.wait()
+        mongoc_change_stream_destroy(changeStream)
+        self.state = .closed
+        self.pollingSemaphore.signal()
+    }
+
     /// Returns the next `T` in the change stream or nil if there is no next value. Will block for a maximum of
     /// `maxAwaitTimeMS` milliseconds as specified in the `ChangeStreamOptions`, or for the server default timeout
     /// if omitted.
-    public func next() -> T? {
-        // We already closed the mongoc change stream, either because we reached the end or encountered an error.
-        guard case let .open(_, connection, client, session) = self.state else {
-            self.error = ClosedChangeStreamError
-            return nil
+    public func next() -> EventLoopFuture<T?> {
+        let operation = NextOperation(target: .changeStream(self))
+        return self.client.executeOperationAsync(operation)
+    }
+
+    /// Returns all of the events seen by this change stream so far.
+    public func all() -> EventLoopFuture<[T]> {
+        let operation = AllOperation(target: .changeStream(self))
+        return self.client.executeOperationAsync(operation)
+    }
+
+    /// Executes the provided closure against each event seen by the change stream.
+    /// This will cause the stream to continuously poll for new events in the background. Once this method has been
+    /// called, other methods that check for events (e.g. all, next, forEach) should not be used.
+    public func forEach(body: @escaping (Result<T, Error>) -> Void) {
+        self.pollingSemaphore.wait()
```

Comment:
usage of this semaphore allows the user to close the change stream even if `forEach` is running in the background.